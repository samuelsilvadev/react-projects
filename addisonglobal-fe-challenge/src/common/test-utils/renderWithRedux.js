import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';

export const renderWithRedux = ({ initialState, middlewares = [] }) => (
	component
) => {
	const mockStore = configureStore(middlewares);
	const store = mockStore(initialState);

	return render(<Provider store={store}>{component}</Provider>);
};
