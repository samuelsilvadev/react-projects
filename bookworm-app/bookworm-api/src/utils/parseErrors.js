module.exports = function parseErrors(errors) {
	const result = {};
	if (errors) {
		Object
			.entries(errors)
			.forEach((value) => {
				result[value[0]] = value[1].message;
			});
	}
	return result;
};
