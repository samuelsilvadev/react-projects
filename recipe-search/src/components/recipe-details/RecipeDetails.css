.details {
	display: flow-root;
	background-color: #f8f9fa;
	min-height: 100vh;
}

.details__wrapper {
	max-width: 700px;
	margin: 110px auto 0;
}

.details__image {
	object-fit: cover;
	max-width: 100%;
	width: 100%;
}

.details__title {
	font-size: 24px;
	font-weight: 700;
	text-transform: uppercase;
	color: #d63031;
}

.details__subtitle {
	font-size: 18px;
	text-transform: uppercase;
}

.details__website {
	font-weight: 400;
}

.details__image,
.details__title,
.details__subtitle,
.details__website {
	margin-bottom: 20px;
}

.details__link {
	align-items: center;
	box-shadow: 0 1px 3px 0 #d4d4d5;
	cursor: pointer;
	display: inline-flex;
	height: 30px;
	justify-content: flex-start;
	padding: 10px;
	text-decoration: none;
}
