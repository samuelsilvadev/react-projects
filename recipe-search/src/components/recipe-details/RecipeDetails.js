import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { recipe } from '@api/recipe';

import './RecipeDetails.css';

export const RecipeDetails = ({ location = {} }) => {
	const [activeRecipe, setActiveRecipe] = useState();

	useEffect(() => {
		const { state } = location;
		if (!state || !state.recipe || !state.recipe.title) {
			if (location.pathname) {
				const recipeId = location.pathname.match(/\d+$/)[0];
				recipe.getById({ recipeId })
					.then((response) => {
						setActiveRecipe(response.recipe);
					});
			}
		} else {
			setActiveRecipe(state.recipe);
		}
	}, [location])

	if (!activeRecipe) {
		return null;
	}

	return (
		<section className="details" data-test="details">
			<div className="details__wrapper">
				<img className="details__image" src={activeRecipe.image_url} alt={recipe.title} data-test="details__image" />
				<h1 className="details__title" data-test="details__title">{activeRecipe.title}</h1>
				<h2 className="details__subtitle" data-test="details__subtitle">Publisher: <span>{activeRecipe.publisher}</span></h2>
				<p className="details__website" data-test="details__website">Website:
					<a href={activeRecipe.publisher_url} rel="noopener noreferrer" target="_blank" data-test="details__website__link">
						{activeRecipe.publisher_url}
					</a>
				</p>
				<Link className="details__link" to="/" data-test="details__link">Go home</Link>
			</div>
		</section>
	);
};

RecipeDetails.propTypes = {
	location: PropTypes.shape({
		pathname: PropTypes.string,
		state: PropTypes.object,
	})
};

export default RecipeDetails;
