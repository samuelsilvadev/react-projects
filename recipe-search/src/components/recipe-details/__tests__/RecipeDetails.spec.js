import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import RecipeDetails from '../RecipeDetails';

import { toTestId } from '@tests/utils';

jest.mock('@api/recipe', () => ({
	recipe: {
		getById: () => Promise.resolve({
			recipe: {
				title: 'Pasta',
				image_url: 'pasta.png',
				publisher: 'Joao',
				publisher_url: 'joao.com'
			}
		})
	}
}));

jest.mock('react-router-dom', () => ({
	Link: ({ to, children, ...rest }) => <a href={to} {...rest}>{children}</a>
}));

const locationMock = {
	pathname: 'recipes/123456'
};

const locationWithStateMock = {
	...locationMock,
	state: {
		recipe: {
			title: 'Pasta',
			image_url: 'pasta.png',
			publisher: 'Joao',
			publisher_url: 'joao.com'
		}
	}
}

const DATA_TESTS = [
	'details',
	'details__image',
	'details__title',
	'details__subtitle',
	'details__website',
	'details__website__link',
	'details__link'
];

describe('<RecipeDetails />', () => {
	it('should render correctly when HAS NO state in location', async () => {
		let wrapper;

		await act(async () => {
			wrapper = mount(<RecipeDetails location={locationMock} />);
		});

		wrapper.update();

		DATA_TESTS.forEach(dataTest => {
			expect(wrapper.find(toTestId(dataTest)).exists()).toBe(true);
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('should render correctly when HAS state in location', async () => {
		let wrapper;

		await act(async () => {
			wrapper = mount(<RecipeDetails location={locationWithStateMock} />);
		});

		wrapper.update();

		DATA_TESTS.forEach(dataTest => {
			expect(wrapper.find(toTestId(dataTest)).exists()).toBe(true);
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('should not render when has no location', async () => {
		let wrapper;

		await act(async () => {
			wrapper = mount(<RecipeDetails />);
		});

		wrapper.update();

		DATA_TESTS.forEach(dataTest => {
			expect(wrapper.find(toTestId(dataTest)).exists()).toBe(false);
		});

		expect(wrapper).toMatchSnapshot();
	});
});
