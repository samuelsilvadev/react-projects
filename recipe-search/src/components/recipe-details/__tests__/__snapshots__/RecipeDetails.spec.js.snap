// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeDetails /> should not render when has no location 1`] = `<RecipeDetails />`;

exports[`<RecipeDetails /> should render correctly when HAS NO state in location 1`] = `
<RecipeDetails
  location={
    Object {
      "pathname": "recipes/123456",
    }
  }
>
  <section
    className="details"
    data-test="details"
  >
    <div
      className="details__wrapper"
    >
      <img
        className="details__image"
        data-test="details__image"
        src="pasta.png"
      />
      <h1
        className="details__title"
        data-test="details__title"
      >
        Pasta
      </h1>
      <h2
        className="details__subtitle"
        data-test="details__subtitle"
      >
        Publisher: 
        <span>
          Joao
        </span>
      </h2>
      <p
        className="details__website"
        data-test="details__website"
      >
        Website:
        <a
          data-test="details__website__link"
          href="joao.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          joao.com
        </a>
      </p>
      <Link
        className="details__link"
        data-test="details__link"
        to="/"
      >
        <a
          className="details__link"
          data-test="details__link"
          href="/"
        >
          Go home
        </a>
      </Link>
    </div>
  </section>
</RecipeDetails>
`;

exports[`<RecipeDetails /> should render correctly when HAS state in location 1`] = `
<RecipeDetails
  location={
    Object {
      "pathname": "recipes/123456",
      "state": Object {
        "recipe": Object {
          "image_url": "pasta.png",
          "publisher": "Joao",
          "publisher_url": "joao.com",
          "title": "Pasta",
        },
      },
    }
  }
>
  <section
    className="details"
    data-test="details"
  >
    <div
      className="details__wrapper"
    >
      <img
        className="details__image"
        data-test="details__image"
        src="pasta.png"
      />
      <h1
        className="details__title"
        data-test="details__title"
      >
        Pasta
      </h1>
      <h2
        className="details__subtitle"
        data-test="details__subtitle"
      >
        Publisher: 
        <span>
          Joao
        </span>
      </h2>
      <p
        className="details__website"
        data-test="details__website"
      >
        Website:
        <a
          data-test="details__website__link"
          href="joao.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          joao.com
        </a>
      </p>
      <Link
        className="details__link"
        data-test="details__link"
        to="/"
      >
        <a
          className="details__link"
          data-test="details__link"
          href="/"
        >
          Go home
        </a>
      </Link>
    </div>
  </section>
</RecipeDetails>
`;
